src/OpenTelemetry/Metrics/CounterMetricSdk.cs

```diff
@@ -39,21 +43,105 @@ public CounterMetricSdk(string name) : this()
+        {
+            // user not using bound instrument. Hence create a
+            // short-lived bound intrument.
+            this.Bind(new LabelSetSdk(labels), isShortLived: true).Add(context, value);
+        }
+
         public override BoundCounterMetric<T> Bind(LabelSet labelset)
+        {
+            // user making Bind call means record is not shortlived.
+            return this.Bind(labelset, isShortLived: false);
+        }
+
+        public override BoundCounterMetric<T> Bind(IEnumerable<KeyValuePair<string, string>> labels)
+        {
+            // user making Bind call means record is not shortlived.
+            return this.Bind(new LabelSetSdk(labels), isShortLived: false);
+        }
+
+        internal BoundCounterMetric<T> Bind(LabelSet labelset, bool isShortLived)
         {
             if (!this.counterBoundInstruments.TryGetValue(labelset, out var boundInstrument))
             {
-                boundInstrument = new BoundCounterMetricSdk<T>();
-
+                var recStatus = isShortLived ? RecordStatus.UpdatePending : RecordStatus.Bound;
+                boundInstrument = new BoundCounterMetricSdk<T>(recStatus);
                 this.counterBoundInstruments.Add(labelset, boundInstrument);
             }
 
+            // if boundInstrument is marked for removal, then take the
+            // lock to sync with Unbind() and re-add. As Collect() might have called Unbind().
+            if (boundInstrument.Status == RecordStatus.CandidateForRemoval)
+            {
+                /*
+                 * If Unbind gets the lock first, then it'd have removed the record.
+                 * But it gets added again by Bind() so no record is lost.
+                 * If Bind method gets this lock first, it'd promote record to UpdatePending, so that
+                 * Unbind will leave this record untouched.
+                                      
+                 * Additional notes:
+                 * This lock is never taken for bound instruments, and they offer the fastest performance.
+                 * This lock is only taken for those labelsets which are marked CandidateForRemoval.
+                 * It means the the 1st time a labelset is re-encountered after a Collect() phase,
+                 * this lock must be taken. Subsequent usage of this labelset before the next Collect()
+                 * will already have status promoted to UpdatePending, and no lock is taken.
+                 * In effect, there is a lock issue after every Collect(). If collect frequency is very high,
+                 * this can affect overall performance.
+                 * Its important to note that, for a brand new LabelSet being encountered for the 1st time, lock is not
+                 * taken. Lock is taken only during the 1st re-appearance of a LabelSet after a Collect period.
+                 *  
+                */
+
+                lock (this.bindUnbindLock)
```

Comment:
Alternatively, we can introduce another status called NoPendingUpdates which stands between
                UpdatePending and CandidateForRemoval.
                With this, the lock is never going to be needed, as long the LabelSet was seen atleast once during the
                Collect interval. Lock will be used only for those rare label sets which occur at a frequency of
                less than a Collect interval. Since lock is only taken for rarely re-occuring label sets, perf
                impact is minimized.

Reply:
I think a state between pending update and candidate for removal makes sense.

Reply:
Agree. Making that change now.